File: Lumo/Models/Rank.swift
import SwiftUI

/// Sıralamadaki rütbeler.
///
/// Elli satırlık düz bir liste bir şey anlatmıyor: 41. olmakla 39. olmak
/// arasında bir his farkı yok. Rütbe onu veriyor — hangi kümede olduğunu
/// görüyorsun, bir üstündeki kümeye ne kadar kaldığını da.
///
/// Eşikler tema değil: renkleri sabit. Bir rütbe herkeste aynı görünmeli,
/// yoksa "mor olan iyidir" gibi ortak bir dil kurulamaz.
enum Rank: Int, CaseIterable, Identifiable {
    case spark = 0, ember, beacon, pulsar, quasar, zenith

    var id: Int { rawValue }

    /// Sonsuz modda bu rütbeye giriş skoru.
    ///
    /// Eşikler doldurmaların tavanına göre ayarlı: en iyi doldurma 33 atlıyor,
    /// yani Pulsar'a ancak değiyor. Kuazar ve Zirve'ye yalnızca gerçek insanlar
    /// çıkabiliyor — en üst iki basamak uydurma bir isimle dolmasın.
    static let endlessFloor: [Int] = [1, 8, 16, 28, 45, 70]

    /// Hız turunda bu rütbenin ÜST süre sınırı (saniye) — altında kalan girer.
    /// İlk eleman sonsuz: en alt rütbeye herkes girer.
    ///
    /// Eşikler baştan çok iyimserdi: Zirve 50 saniye istiyordu, oysa on bölümün
    /// gerçekte inebildiği yer 1:59 civarı. Üst üç basamak kimsenin
    /// giremeyeceği bir yerdeydi, yani hiç yokmuş gibiydi. Artık Zirve 2:05'in
    /// altı — zor ama insan işi.
    static let speedrunCeiling: [Double] = [.infinity, 240, 200, 170, 145, 125]

    /// Adlar bilerek ışığın şiddetine göre: kıvılcımdan zirveye. Hiçbiri
    /// küre stili ya da tema adıyla çakışmıyor (Nova, Comet, Aurora hepsi
    /// başka şeylerin adı) — biri karakter, biri arka plan, biri rütbe;
    /// üçü karışsa oyuncu neyi kazandığını anlamaz.
    var title: LocalizedStringKey {
        switch self {
        case .spark:  return "Spark"
        case .ember:  return "Ember"
        case .beacon: return "Beacon"
        case .pulsar: return "Pulsar"
        case .quasar: return "Quasar"
        case .zenith: return "Zenith"
        }
    }

    /// Griden turuncuya, altına, camgöbeğine, mora, en üstte beyaza.
    /// Sıcaklık değil parlaklık artıyor — yukarı çıkmak "daha aydınlık".
    var color: Color {
        switch self {
        case .spark:  return Color(red: 0.54, green: 0.58, blue: 0.65)
        case .ember:  return Color(red: 0.91, green: 0.53, blue: 0.24)
        case .beacon: return Color(red: 0.94, green: 0.77, blue: 0.23)
        case .pulsar: return Color(red: 0.27, green: 0.78, blue: 0.88)
        case .quasar: return Color(red: 0.64, green: 0.42, blue: 0.96)
        case .zenith: return Color(red: 0.92, green: 0.95, blue: 1.00)
        }
    }

    /// Bir sonucun rütbesi
    static func of(value: Double, mode: LeaderboardMode) -> Rank {
        switch mode {
        case .endless:
            var found = Rank.spark
            for rank in allCases where Int(value) >= endlessFloor[rank.rawValue] { found = rank }
            return found
        case .speedrun:
            var found = Rank.spark
            for rank in allCases where value < speedrunCeiling[rank.rawValue] { found = rank }
            return found
        }
    }

    /// Rütbe aralığının okunur hâli — cetvelde gösteriliyor
    func rangeText(mode: LeaderboardMode) -> String {
        switch mode {
        case .endless:
            let low = Self.endlessFloor[rawValue]
            guard rawValue < Self.allCases.count - 1 else { return "\(low)+" }
            return "\(low)–\(Self.endlessFloor[rawValue + 1] - 1)"
        case .speedrun:
            let f = Self.clock
            guard rawValue > 0 else { return "\(f(Self.speedrunCeiling[1]))+" }
            guard rawValue < Self.allCases.count - 1 else { return "< \(f(Self.speedrunCeiling[rawValue]))" }
            return "\(f(Self.speedrunCeiling[rawValue + 1]))–\(f(Self.speedrunCeiling[rawValue]))"
        }
    }

    /// Cetveldeki süreler salise istemiyor: "1:45", "1:45.00" değil
    private static func clock(_ t: Double) -> String {
        String(format: "%d:%02d", Int(t) / 60, Int(t) % 60)
    }
}

/// Rütbe nişanı: oyunun kendi dilinde küçük bir küre. Simge yerine küre,
/// çünkü ekranın geri kalanı zaten kürelerden oluşuyor.
struct RankBadge: View {
    let rank: Rank
    var size: CGFloat = 16

    var body: some View {
        Circle()
            .fill(
                RadialGradient(colors: [.white.opacity(0.9), rank.color],
                               center: UnitPoint(x: 0.35, y: 0.3),
                               startRadius: 0, endRadius: size * 0.7)
            )
            .frame(width: size, height: size)
            .overlay(Circle().strokeBorder(.white.opacity(0.25), lineWidth: 0.5))
            .shadow(color: rank.color.opacity(0.8), radius: size * 0.3)
    }
}

File: android/app/src/main/java/com/caganhatapci/orbeon/model/Rank.kt
package com.caganhatapci.orbeon.model

import androidx.compose.ui.graphics.Color
import com.caganhatapci.orbeon.services.LeaderboardService

/**
 * Sıralamadaki rütbeler — iOS'takiyle birebir aynı eşikler ve renkler.
 *
 * Elli satırlık düz bir liste bir şey anlatmıyor: 41. olmakla 39. olmak
 * arasında bir his farkı yok. Rütbe onu veriyor — hangi kümede olduğunu
 * görüyorsun, bir üstündeki kümeye ne kadar kaldığını da.
 *
 * Renkler TEMAYA BAĞLI DEĞİL, sabit. Bir rütbe herkeste aynı görünmeli,
 * yoksa "mor olan iyidir" gibi ortak bir dil kurulamaz.
 */
enum class Rank(val titleRes: Int, val color: Color) {
    // Griden turuncuya, altına, camgöbeğine, mora, en üstte beyaza.
    // Sıcaklık değil PARLAKLIK artıyor — yukarı çıkmak "daha aydınlık".
    SPARK(com.caganhatapci.orbeon.R.string.rank_spark, Color(0xFF8A94A6)),
    EMBER(com.caganhatapci.orbeon.R.string.rank_ember, Color(0xFFE8873D)),
    BEACON(com.caganhatapci.orbeon.R.string.rank_beacon, Color(0xFFF0C43B)),
    PULSAR(com.caganhatapci.orbeon.R.string.rank_pulsar, Color(0xFF45C7E0)),
    QUASAR(com.caganhatapci.orbeon.R.string.rank_quasar, Color(0xFFA36BF5)),
    ZENITH(com.caganhatapci.orbeon.R.string.rank_zenith, Color(0xFFEBF2FF));

    companion object {
        /**
         * Sonsuz modda bu rütbeye giriş skoru.
         *
         * Eşikler doldurmaların tavanına göre ayarlı: en iyi doldurma 33
         * atlıyor, yani Pulsar'a ancak değiyor. Kuazar ve Zirve'ye yalnızca
         * gerçek insanlar çıkabiliyor — en üst iki basamak uydurma bir isimle
         * dolmasın.
         */
        val ENDLESS_FLOOR = intArrayOf(1, 8, 16, 28, 45, 70)

        /**
         * Hız turunda bu rütbenin ÜST süre sınırı (saniye) — altında kalan
         * girer. İlk eleman sonsuz: en alt rütbeye herkes girer.
         *
         * Eşikler baştan çok iyimserdi: Zirve 50 saniye istiyordu, oysa on
         * bölümün gerçekte inebildiği yer 1:59 civarı.
         */
        val SPEEDRUN_CEILING = doubleArrayOf(Double.MAX_VALUE, 240.0, 200.0, 170.0, 145.0, 125.0)

        /** Bir sonucun rütbesi */
        fun of(value: Double, mode: LeaderboardService.Mode): Rank {
            var found = SPARK
            if (mode == LeaderboardService.Mode.ENDLESS) {
                for (r in entries) if (value.toInt() >= ENDLESS_FLOOR[r.ordinal]) found = r
            } else {
                for (r in entries) if (value < SPEEDRUN_CEILING[r.ordinal]) found = r
            }
            return found
        }

        private fun clock(t: Double): String =
            String.format("%d:%02d", t.toInt() / 60, t.toInt() % 60)

        /** Rütbe aralığının okunur hâli — cetvelde gösteriliyor */
        fun rangeText(rank: Rank, mode: LeaderboardService.Mode): String {
            val i = rank.ordinal
            return if (mode == LeaderboardService.Mode.ENDLESS) {
                val low = ENDLESS_FLOOR[i]
                if (i == entries.size - 1) "$low+" else "$low–${ENDLESS_FLOOR[i + 1] - 1}"
            } else {
                when {
                    i == 0 -> "${clock(SPEEDRUN_CEILING[1])}+"
                    i == entries.size - 1 -> "< ${clock(SPEEDRUN_CEILING[i])}"
                    else -> "${clock(SPEEDRUN_CEILING[i + 1])}–${clock(SPEEDRUN_CEILING[i])}"
                }
            }
        }
    }
}
